### PR TITLE
Fixes #28374 - disable run ansible roles option from hostgroup dropdown if no roles are assigned

### PR DESCRIPTION
**Pull request by nofaralfasi**
Remove `Run all Ansible roles` option from hostgroup dropdown if no roles are assigned.

### CODE REVIEW

File: app/helpers/foreman_ansible/ansible_hostgroups_helper.rb [hunk 1]
```diff
@@ -3,13 +3,9 @@
...
     def ansible_hostgroups_actions(hostgroup)
-      play_roles = if hostgroup.all_ansible_roles.empty?
-                     { action: (link_to _('Run all Ansible roles'), 'javascript:void(0);', disabled: true, title: 'No Roles assigned'), priority: 31 }
-                   else
-                     { action: display_link_if_authorized(_('Run all Ansible roles'), hash_for_play_roles_hostgroup_path(id: hostgroup), 'data-no-turbolink': true, title: _('Run all Ansible roles on hosts belonging to this host group')), priority: 31 }
-                   end
+      return [] if hostgroup.all_ansible_roles.empty? || !User.current.can?(:create_job_invocations)
 
-      [play_roles] if User.current.can?(:create_job_invocations)
+      [{ action: display_link_if_authorized(_('Run all Ansible roles'), hash_for_play_roles_hostgroup_path(id: hostgroup), 'data-no-turbolink': true, title: _('Run all Ansible roles on hosts belonging to this host group')), priority: 31 }]
```
Comment (stejskalleos) [inline app/helpers/foreman_ansible/ansible_hostgroups_helper.rb:8]: I like the code and it works as expected, however I'm not sure if removing the link is correct approach.
As a user, I find super confusing to have missing links just for some groups, maybe disabled link with hover help text explaining why it's disabled is better user experience.
I'm not UX guy, I would suggest to contact Maria Svirikova, she's the person to go for UX, she will know what to do.

Comment (nofaralfasi) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb:8]: You are right, I'll contact Maria and ask her.

Comment (nofaralfasi) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb:8]: Hi @MariSvirik, we want to make a UI change and your opinion could help us here.

On the Hostgroup page, we have the `Run all Ansible roles` option in the actions dropdown. 
However, not all Hostgroups have ansible roles assigned to them.
Currently if you hover over this option for a Hostgroup without any Ansible roles you'll see this tooltip:
image

We are thinking about disabling or removing the `Run all Ansible roles` option from the dropdown for Hostgroups without any Ansible roles.

What do you think is the best approach here?

Comment (MariSvirik) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb:8]: Hi @nofaralfasi ,  @stejskalleos is right, the good practice is to disable the action and add a tooltip with an explanation. If the action could be enabled by users/their actions (which in this case is) and makes sense in the context of the page, this is the preferable approach.

Comment (nofaralfasi) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb:8]: Great, thank you @MariSvirik! 
I will apply the new changes and check with you after to make sure it is working properly.

Comment (nofaralfasi) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb:8]: Hi @MariSvirik, this is how the disabled action looks like with the new changes:
Screenshot from 2022-08-23 17-45-28

Let me know if something is missing or needs to be changed.
Thanks!

**Comment by nofaralfasi** [comment]
This PR is blocked by https://github.com/theforeman/foreman/pull/9367

File: app/helpers/foreman_ansible/ansible_hostgroups_helper.rb (hunk 1 above)
Comment (MariSvirik) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb]: thank you for pinging me @nofaralfasi 
1. Remove the ban icon
2.  text should be either "No roles assigned" (sentence case) OR "To run all Ansible roles, firstly assign roles to the hostgroup." (where you explain how to enable the action)

- phrasing can be checked by the doc team (pinging them now)

Comment (nofaralfasi) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb]: Great, I'll do the necessary changes.
Thank you for your help!

Comment (nofaralfasi) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb]: Hi @MariSvirik, is there an update from the doc team regarding the phrasing?
Thanks!

**Comment by AkshayGadhaveRH** [comment]
Hey @nofaralfasi @MariSvirik, does `No roles assigned, assign roles to the hostgroup first` sound better? Since the option already says `Run all Ansible Roles`, we might not have to repeat it. IMO the two important bits of information here are that there are no roles assigned, and how to assign them. What do you guys think?

**Comment by nofaralfasi** [comment]
> Hey @nofaralfasi @MariSvirik, does `No roles assigned, assign roles to the hostgroup first` sound better? Since the option already says `Run all Ansible Roles`, we might not have to repeat it. IMO the two important bits of information here are that there are no roles assigned, and how to assign them. What do you guys think?

Hi @AkshayGadhave, thank you for your response.
I'm no expert, but I also agree with you, `No roles assigned, assign roles to the hostgroup first` sounds good. 
@MariSvirik let us know what you think :)

File: .github/workflows/ruby_tests.yml [hunk 2]
```diff
@@ -66,6 +66,8 @@ jobs:
           bundle exec rake db:create
           bundle exec rake db:migrate
           bundle exec rake db:test:prepare
+          npm install
```
Comment (Ron-Lavi) [inline .github/workflows/ruby_tests.yml:69]: Can you move it to line 63 (under the Node setup):
```yml
        - name: Setup Node
        if: github.event_name != 'push'
        uses: actions/setup-node@v1
        with:
          node-version:  ${{ matrix.node-version }}
        - name: Install npm dependencies
        run: npm install
```

File: .github/workflows/ruby_tests.yml [hunk 3]
```diff
@@ -66,6 +66,8 @@ jobs:
...
+          bundle exec rake webpack:compile
```
[the remark this answers is not in the document]
Comment (Ron-Lavi) [reply .github/workflows/ruby_tests.yml:70]: never mind, it should work as well

**Comment by stejskalleos** [comment]
I see that https://github.com/theforeman/foreman/pull/9367 is ready for review, is this PR ready for review too?

**Comment by nofaralfasi** [comment]
> I see that [theforeman/foreman#9367](https://github.com/theforeman/foreman/pull/9367) is ready for review, is this PR ready for review too?

It is, but I still have some tests that are failing.

File: .github/workflows/ruby_tests.yml [hunk 4]
```diff
@@ -60,12 +60,17 @@ jobs:
         uses: actions/setup-node@v1
         with:
           node-version:  ${{ matrix.node-version }}
+      - name: Install npm dependencies
+        run: npm install
```
Comment (Ron-Lavi) [inline .github/workflows/ruby_tests.yml:64]: can you also add an additional step to install the plugin's dependencies?
e.g:
```yml
- name: Install plugin's npm dependencies
  run: npm install --omit=dev
  working-directory: ${{ github.workspace }}/foreman_ansible
```

**Review by Ron-Lavi — COMMENTED** [review]
Since this feature will be based on foreman v3.5,
can you also update https://github.com/theforeman/foreman_ansible/blob/master/lib/foreman_ansible/register.rb#L4
and https://github.com/theforeman/foreman_ansible/blob/master/lib/foreman_ansible/version.rb#L7 to `10.0.0`

File: lib/foreman_ansible/version.rb [hunk 5]
```diff
@@ -4,5 +4,5 @@
 # This way other parts of Foreman can just call ForemanAnsible::VERSION
 # and detect what version the plugin is running.
 module ForemanAnsible
-  VERSION = '9.0.1'
+  VERSION = '10.0.0'
```
Comment (stejskalleos) [inline lib/foreman_ansible/version.rb:7]: @Ron-Lavi Shouldn't we change the version just before releasing it?

File: app/helpers/foreman_ansible/ansible_hostgroups_helper.rb (hunk 1 above)
Comment (nofaralfasi) [reply app/helpers/foreman_ansible/ansible_hostgroups_helper.rb]: @MariSvirik 
Currently I only updated the text to sentence case "No roles assigned". If the text needs to be changed, please let me know. 
Thanks.

File: lib/foreman_ansible/version.rb (hunk 5 above)
Comment (Ron-Lavi) [reply lib/foreman_ansible/version.rb:7]: imho it will be more organized to do it once we touch the required foreman version

Comment (stejskalleos) [reply lib/foreman_ansible/version.rb:7]: Yop, I'm okay with that if we keep the change in separate commit, so it's easily found in the log

**Comment by Ron-Lavi** [comment]
Now tests are failing for different reasons :)

**Comment by nofaralfasi** [comment]
> Now tests are failing for different reasons :)

Yeah, I will first fix the tests and then commit the version update.

File: package.json [hunk 6]
```diff
@@ -21,7 +21,6 @@
     "@theforeman/stories": "^10.0",
     "@theforeman/test": "^10.0",
     "@theforeman/vendor-dev": "^10.0",
-    "@testing-library/user-event": "^13.2.1",
```
Comment (stejskalleos) [inline package.json:24]: Is this intentional?

Comment (nofaralfasi) [reply package.json:24]: I can try to delete it and see if it helps.

Comment (Ron-Lavi) [reply package.json:24]: I guess we need this package for JS tests

**Comment by Ron-Lavi** [comment]
it's ok that tests are failing, we need https://github.com/theforeman/foreman/pull/9367 to be merged and then retest it

**Comment by Ron-Lavi** [comment]
Upstream PR got merged, I triggered the test again and now it fails with something else

File: package.json (hunk 6 above)
Comment (nofaralfasi) [reply package.json:24]: It's written there twice (also on [line 17](https://github.com/theforeman/foreman_ansible/blob/ba8db81538f635f129f41b8a217736ee107c3664/package.json#L17)).

**Comment by nofaralfasi** [comment]
> The Foreman PR got merged, I triggered the test again and now it fails with something else :)
> 
> EDIT: Still failing on the `HostgroupJsTest` and also on `ActiveRecord::AssociationTypeMismatch: RemoteExecutionFeature`
> 
> @adamruzicka do we need to change anything in tests?
> 
> ```
> Error:
> HostgroupJsTest#test_0001_hostgroup without roles should have disabled link:
> Capybara::ExpectationNotMet: expected to find link "Run all Ansible roles" at least 1 time but there were no matches
>     foreman_ansible/test/integration/hostgroup_js_test.rb:19:in `block in <class:HostgroupJsTest>'
> 
> Error:  
> AnsibleProviderTest::.proxy_command_options::when it is using the ansible_run_host feature#test_0001_has remote_execution_command false:
> ActiveRecord::AssociationTypeMismatch: RemoteExecutionFeature(#72560) expected, got nil which is an instance of NilClass(#40)
>     foreman_ansible/test/unit/ansible_provider_test.rb:27:in `block (3 levels) in <class:AnsibleProviderTest>'
> ```

Hi @Ron-Lavi, thank you for the help, and for re-running the tests.
I don't see the first error you mentioned anymore (`HostgroupJsTest#test_0001_hostgroup without roles should have disabled link`).
And the second error is not related to my changes, so I am curious to know what caused it.

**Review by Ron-Lavi — COMMENTED** [review]
> And the second error is not related to my changes, so I am curious to know what caused it.

It might be related to the updated required_foreman version,
can you try updating 
https://github.com/theforeman/foreman_ansible/blob/7b41bdcc8494e3146c8c7f767e4a6fbc7d4d564d/foreman_ansible.gemspec#L23:L24

to:
```rb
  s.add_dependency 'foreman_remote_execution', '>= 8.0.0'
  s.add_dependency 'foreman-tasks', '>= 7.0.0'
```

**Comment by Ron-Lavi** [comment]
@nofaralfasi can you change the commit message to `Bump version 10.0.0` plus the explanation as a message?

**Review by Ron-Lavi — APPROVED** [review]
Thanks @nofaralfasi 🏆

**Pull request merged**